Marlin/src/gcode/calibrate/G425.cpp

```diff
@@ -0,0 +1,673 @@
+ *
+ * Based on Sprinter and grbl.
+ * Copyright (C) 2011 Camiel Gubbels / Erik van der Zalm
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#include "../../Marlin.h"
+
+#if ENABLED(CALIBRATION_GCODE)
+
+#include "../gcode.h"
+#include "../../lcd/ultralcd.h"
+#include "../../module/motion.h"
+#include "../../module/planner.h"
+#include "../../module/tool_change.h"
+#include "../../module/endstops.h"
+
+//#define CALIBRATION_MEASUREMENT_DEBUG
+
+/**
+ * G425 backs away from the cube by various distances depending
+ * on the user's confidence level:
+ *
+ *   UNKNOWN   - No real notion on where the cube is on the bed
+ *   UNCERTAIN - Measurement may be uncertain due to backlash
+ *   CERTAIN   - Measurement obtained with backlash compensation
+ */
+
+#ifndef CALIBRATION_MEASUREMENT_UNKNOWN
+  #define CALIBRATION_MEASUREMENT_UNKNOWN   5.0 // mm
+#endif
+#ifndef CALIBRATION_MEASUREMENT_UNCERTAIN
+  #define CALIBRATION_MEASUREMENT_UNCERTAIN 1.0 // mm
+#endif
+#ifndef CALIBRATION_MEASUREMENT_CERTAIN
+  #define CALIBRATION_MEASUREMENT_CERTAIN   0.5 // mm
+#endif
+
+#define HAS_X_CENTER (ENABLED(CALIBRATION_CUBE_MEASURE_LEFT)  && ENABLED(CALIBRATION_CUBE_MEASURE_RIGHT))
+#define HAS_Y_CENTER (ENABLED(CALIBRATION_CUBE_MEASURE_FRONT) && ENABLED(CALIBRATION_CUBE_MEASURE_BACK))
+
+#if ENABLED(BACKLASH_GCODE)
+  extern float backlash_distance_mm[], backlash_correction, backlash_smoothing_mm;
+#endif
+
+enum side_t : uint8_t { TOP, RIGHT, FRONT, LEFT, BACK, NUM_SIDES };
+
+struct measurements_t {
+  static const float dimensions[XYZ];
+  static const float true_center[XYZ]; // This cannot be constexpr since it is accessed by index in probe_cube_side
+  float cube_center[XYZ] = CALIBRATION_CUBE_CENTER;
+  float cube_side[NUM_SIDES];
+
+  float backlash[NUM_SIDES];
+  float pos_error[XYZ];
+
+  float nozzle_outer_dimension[2] = {CALIBRATION_NOZZLE_OUTER_DIAMETER, CALIBRATION_NOZZLE_OUTER_DIAMETER};
+};
+
+const float measurements_t::true_center[XYZ] = CALIBRATION_CUBE_CENTER;
+
+const float measurements_t::dimensions[]  = CALIBRATION_CUBE_DIMENSIONS;
+
+/**
+ * A class to save and change the endstop state,
+ * then restore it when it goes out of scope.
+ */
+class TemporaryGlobalEndstopsState {
+  bool saved;
+
+  public:
+    TemporaryGlobalEndstopsState(bool enable) : saved(endstops.are_endstops_enabled_globally()) {
```

Comment:
Issue here on 88
`TemporaryGlobalEndstopsState(bool enable) : saved(endstops.are_endstops_enabled_globally()) {`
should be
`TemporaryGlobalEndstopsState(bool enable) : saved(endstops.global_enabled()) {`

Reply:
@SJ-Innovation: Ah. I think it is endstops.are_endstops_enabled_globally() on my FW. I guess I need to update my code!

Reply:
Its definitely `global_enabled` in the latest versions... An update might be helpful :)